algolia/search/requests_search.go

```diff
@@ -1,18 +1,79 @@
 package search
 
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/algolia/algoliasearch-client-go/algolia/internal/opt"
+)
+
 type searchReq struct {
 	Params string `json:"params"`
 }
 
 type searchParams struct {
 	Query string `json:"query"`
 	QueryParams
+	ExtraParams map[string]interface{} `json:"-"`
+}
+
+func (p searchParams) MarshalJSON() ([]byte, error) {
+
+	// First, if there is no extra parameter, the structure is
+	// serialized as-is, using a type alias to avoid infinite
+	// recursion of MarshalJSON calls.
+
+	if len(p.ExtraParams) == 0 {
+		type searchParamsAlias searchParams
+		return json.Marshal(searchParamsAlias(p))
+	}
+
+	// By there, because we do have extra search parameters, it
+	// is needed to flatten them in the same JSON payload, along
+	// with the known ones.
+	//
+	// To do that, we create a new type at runtime using the reflect
+	// package and populate it with:
+	//
+	//   1. the known fields: "query" + any other fields from
+	//      QueryParams if they are present
+	//   2. the unknown fields from the ExtraParams map
```

Comment:
cf my remark regarding key conflict in the `merged` map; I think we should allow to override a param that was set explicitely i.e. 
```
client.Search(query, opt.Analytics(false), opt.ExtraOptions(map[string]interface{}{"analytics": true}))
``` 
should _not_ work in my opinion.

To extend on this, iirc, 
* `client.Search(query, opt.Analytics(false), opt.ExtraOptions(map[string]interface{}{"analytics": true}))` would end up with `analytics=true` 
* `client.Search(query, opt.ExtraOptions(map[string]interface{}{"analytics": true}), opt.Analytics(false))` would end up with `analytics=false`.

Note that I'm not asking to fail because the parameter has an explicit version but because there is a key conflict. Preventing parameters that can be explicitely set would be nice to have imo but maybe overkill.

Reply:
I think both point of views are valid actually. For the `opt.XXX > ExtraOptions(XXX)` that you are suggesting, I see your explanation as a _pro_ but `opt.ExtraURLParams` and `opt.ExtraHeaders` are actually overriding the values of any existing header. Not that those `opt.Extra` are advanced options that no one shoud use.

In short, while I see why your point is valid as wel, again for consistency with other `opt.Extra*` options, I would keep the following behavior: "if you are using `opt.Extra*` options, you know what you are doing, those are advanced options".

WDYT @Ant-hem

Reply:
I'd go for overriding the values with `opt` because overriding with `opt` is the current behavior for in the all the clients.

Reply:
so if we have a consistent behavior across clients let's go for it. 
From a dev perspective, I would expect to have the extra params to typically be generated by code vs explicit options being "hardcoded" so I would find it safer to not have any overriding.

As Antoine said, we should have an explicit test on the overriding behavior!

Reply:
Adn as a follow up, maybe it's already done and you can discard my comment, but this should be really explicitely documented.